Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+        ///  - .failed (if an error was found in the request soundness check)
+        case verifyingRequest
+        /// Waiting for the channel to be writable. Valid transitions are:
+        ///  - .running(.streaming, .initialized) (once the Channel is writable again and if a request body is expected)
+        ///  - .running(.endSent, .initialized) (once the Channel is writable again and no request body is expected)
+        ///  - .failed (if a connection error occurred)
+        case waitForChannelToBecomeWritable(HTTPRequestHead)
+        /// A request is on the wire. Valid transitions are:
+        ///  - .finished
+        ///  - .failed
+        case running(RequestState, ResponseState)
+        /// The request has completed successfully
+        case finished
+        /// The request has failed
+        case failed(Error)
+    }
+
+    /// A sub state for a running request. More specifically for sending a request body.
+    fileprivate enum RequestState {
+        /// A sub state for sending a request body. Stores whether a producer should produce more
+        /// bytes or should pause.
+        enum ProducerControlState: Equatable {
+            /// The request body producer should produce more body bytes. The channel is writable,
+            case producing
+            /// The request body producer should pause producing more bytes. The channel is not writable,
+            case paused
+        }
+
+        /// The request is streaming its request body. `expectedBodyLength` has a value, if the request header contained
+        /// a `"content-length"` header field. It is the request header contained a `"transfer-encoding" = "chunked"`
+        /// header field.
+        case streaming(expectedBodyLength: Int?, sentBodyBytes: Int, producer: ProducerControlState)
+        /// The request has sent its request body and end.
+        case endSent
+    }
+
+    fileprivate enum ResponseState {
+        /// A sub state for receiving a response. Stores whether the consumer has either signaled demand for more data or
+        /// is busy consuming the so far forwarded bytes
+        enum ConsumerControlState {
+            case downstreamIsConsuming(readPending: Bool)
+            case downstreamHasDemand
+        }
+
+        /// A response head has not been received yet.
+        case waitingForHead
+        /// A response head has been received and we are ready to consume more data of the wire
+        case receivingBody(HTTPResponseHead, ConsumerControlState)
+        /// A response end has been received and we are ready to consume more data of the wire
+        case endReceived
+    }
+
+    enum Action {
+        /// A action to execute, when we consider a request "done".
+        enum FinalStreamAction {
+            /// close the connection
+            case close
+            /// trigger a read event
+            case read
+            /// do nothing
+            case none
```

Comment:
It would be useful to clarify when we do `none` here: namely, when we hit an error validating the request and haven't sent it yet.